src/index.js

```diff
@@ -1,126 +1,252 @@
     }
 
-    while ((next = lifecycle.pop())) next()
+    // Flush the lifecycle stack and notify subscribers of created/updated nodes.
+
+    while ((next = lifecycleStack.pop())) next()
   }
 
+  /**
+   * Schedules a new render call if we are not currently locked and toggles the
+   * lock in order to throttle successive repaint attempts, e.g., when calling
+   * several actions in a row.
+   *
+   * @private
+   */
   function repaint() {
-    if (!patchLock) {
-      patchLock = !patchLock
+    if (!renderLock) {
+      renderLock = !renderLock
       setTimeout(render)
     }
   }
 
-  function copy(a, b) {
-    var target = {}
-
-    for (var i in a) target[i] = a[i]
-    for (var i in b) target[i] = b[i]
-
-    return target
+  /**
+   * Copy all properties from source into target immutably.
+   *
+   * @param {Object} target Object to copy into.
+   * @param {Object} source Object to copy from.
+   * @returns {Object}
+   * @private
+   */
+  function copy(target, source) {
+    var obj = {}
+
+    for (var i in target) obj[i] = target[i]
+    for (var i in source) obj[i] = source[i]
+
+    return obj
   }
 
-  function set(path, value, source, target) {
+  /**
+   * Sets the value at path of the source object.
+   *
+   * @private
+   */
+  function set(path, value, target, source) {
     if (path.length) {
       target[path[0]] =
-        path.length > 1 ? set(path.slice(1), value, source[path[0]], {}) : value
+        path.length > 1 ? set(path.slice(1), value, {}, source[path[0]]) : value
       return copy(source, target)
     }
     return value
   }
 
+  /**
+   * Get the property at path of the given object.
+   *
+   * @param {Array} path The path of the property to get.
+   * @param {Object} source The object to query.
+   *
+   * @private
+   */
   function get(path, source) {
     for (var i = 0; i < path.length; i++) {
       source = source[path[i]]
     }
     return source
   }
 
-  function init(path, slice, actions) {
-    for (var key in actions) {
-      typeof actions[key] === "function"
+  /**
+   * Enhance the state (add missing leaves from the actions tree) and actions
+   * to trigger repaints after updating the state.
+   *
+   * @param {Array} path
+   * @param {Object} substate
+   * @param {Object} subactions
+   *
+   * @private
+   */
+  function enhance(path, substate, subactions) {
+    for (var key in subactions) {
+      typeof subactions[key] === "function"
         ? (function(key, action) {
-            actions[key] = function(data) {
+            subactions[key] = function(data) {
               if (typeof (data = action(data)) === "function") {
-                data = data(get(path, state), actions)
+                data = data(get(path, state), subactions)
               }
 
-              if (data && data !== (slice = get(path, state)) && !data.then) {
-                repaint((state = set(path, copy(slice, data), state, {})))
+              if (
+                data &&
```

Comment:
Comment about optimization for returning falsey/same state/`Promise`?

Reply:
Added only this comment `// Promise`.